### DISCUSSION TITLE
Typer/Cython - anyone ever used it?

### DISCUSSION THREAD

**Discussion by theodore86** [fastapi/typer#684] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer

def main(user_name: Optional[List[str]] = typer.Option("Theo", "--user-name", "-n")):
    print(f"Hello {user_name}")
```

### Description

Python 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from test_lib.test_file import main
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src/test_lib/test_file.py", line 31, in init test_lib.test_file
    def main(user_name: Optional[List[str]] = typer.Option("Theo", "--user-name", "-n")):
TypeError: Expected list, got OptionInfo
>>> 

### Operating System

Linux

### Operating System Details

Cython==3.0.4
Ubuntu 22.04

### Typer Version

0.9.0

### Python Version

3.10.6

### Additional Context

Well it seems that i having some issues with code that has been cythonized and typer is being as a CLI, is there any past experience or recommendation on how to fix the above issue?

The annotated syntax is not working (```typing.Annotated[]```) the CLI is getting messed up so i decided to try out  the non-Annotated syntax for the CLI options.

Thanks in advance,
Theo G.

**Comment by theodore86** [comment]
Not sure if anyone ever used Typer with Cython but the ```typing.Annotated[]``` arguments are not working.

**Comment by melsabagh** [comment]
It seems that the problem is in this line: 

https://github.com/tiangolo/typer/blob/3a7264cd56181690805f220cb44a0301c0fdf9f3/typer/utils.py#L109-L109

Cython enables `from __future__ import annotations` by default which results in stringifying the annotations. According to the [docs](https://docs.python.org/3/library/inspect.html#inspect.signature), `inspect.signature(..., eval_str=True)` should be used to unstringify these annotations. This is related to https://github.com/tiangolo/typer/discussions/598.

@tiangolo 

I am currently monkey-patching Typer as follows to workaround this: 

```python
import inspect

import typer.utils

class _inspect:
    @staticmethod
    def signature(func, *args, **kwargs):
        eval_str = kwargs.pop("eval_str", True)
        return inspect.signature(func, *args, **kwargs, eval_str=eval_str)
typer.utils.inspect = _inspect
```

**Reply by theodore86** [reply]
Thank you @melsabagh, i will try to re-use this proposal, it would very useful to have this fix in the next typer release

**Comment by svlandeg** [comment]
Related PR: https://github.com/tiangolo/typer/pull/721

**Reply by tiangolo** [reply, marked as answer]
Thanks for the report @theodore86!

This was indeed (probably) solved by #721, available in Typer 0.9.3, released in the next few hours. :rocket:
